### PR TITLE
[DTensor] Fix `to_local` backward by providing default `grad_placement` type

### LINKED ISSUE #172932: [DTensor] `to_local` + `from_local` on `Partial` DTensor produce different gradient
## Description

When using `DTensor.to_local()` followed by `DTensor.from_local()` on a DTensor with `Partial` placement, the forward pass appears to be a no-op (as expected), but the backward pass produces different gradients compared to using the DTensor directly.

Why aren't the gradients identical when using `to_local()` followed by `from_local()` with the same placements? Shouldn't this pattern be a no-op, resulting in the same gradients as using the DTensor directly?

This design is error-prone in some scenarios. Would it be better to add an extra check in `DTensor.to_local` to raise an error if the placements is `Partial` and `grad_placements` is not `Replicate()`?

## Reproduce Code
```py
# torchrun --nproc_per_node=2 test_grad.py
import torch
import loguru
from torch.distributed import init_process_group
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed._tensor import DTensor, distribute_tensor, Shard

init_process_group()
torch.cuda.set_device(torch.distributed.get_rank())
device_mesh = init_device_mesh(device_type='cuda', mesh_shape=(2,), mesh_dim_names=('tp',))

def test_case(plan=1):
    original_tensor = torch.tensor([1., 2.], device='cuda', requires_grad=True)
    original_tensor = distribute_tensor(original_tensor, device_mesh, [Shard(0)])
    out = original_tensor.sum()
    if plan == 1:
        out = DTensor.from_local(
            out.to_local(), out.device_mesh, out.placements, run_check=False
        )
    elif plan == 2:
        pass # do nothing

    loss = out.full_tensor().sum()
    loss.backward()
    grad = original_tensor.grad.full_tensor()
    return loss, grad

loss1, grad1 = test_case(1)
loss2, grad2 = test_case(2)
loguru.logger.info(f"Plan 1 Loss: {loss1} Grad: {grad1}")
loguru.logger.info(f"Plan 2 Loss: {loss2} Grad: {grad2}")
```

## Output
```
Plan 1 Loss: 3.0 Grad: tensor([2., 2.], device='cuda:1')
Plan 1 Loss: 3.0 Grad: tensor([2., 2.], device='cuda:0')
Plan 2 Loss: 3.0 Grad: tensor([1., 1.], device='cuda:1')
Plan 2 Loss: 3.0 Grad: tensor([1., 1.], device='cuda:0')
```

### Versions

2.9.1

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

### PR DESCRIPTION
**Pull request by acisseJZhong**
Fixes #172932, see comment [here](https://github.com/pytorch/pytorch/issues/172932#issuecomment-3802332176) 

**Problem**: `DTensor.to_local()` followed by `DTensor.from_local()` on `Partial` produces different gradients than pure DTensor operations.

**Root cause:** during from_local backward, we always redistribute gradients to Replicate when forward is Partial. However, this information is [not saved](https://github.com/pytorch/pytorch/blob/635ff65bc517c175360a0a85502488ee341059c1/torch/distributed/tensor/_api.py#L91) in to_local backward, where it take the `Replicated `tensor but still annotate as `Partial`.

**Fix:** Similarly to the fix for `from_local` in https://github.com/pytorch/pytorch/pull/173153, we propose this fix to provide a default `Replicate` gradient for `Partial` input(as we always map forward partial to backward replicate in `from_local`) when `grad_placements` is not provided. However, this fix does not guarantee correctness and might break BC. To guarantee correctness, users are  **strongly recommended** to pass in `grad_placement` type into `to_local`.

cc @KimbingNg (thanks for the test case!)

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# torchrun --nproc_per_node=2 test.py
```
Comment (wconstab) [inline test.py]: did you mean to put this as an actual test in test/distributed/..?

File: test.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+
+def test_case(plan=1):
+    original_tensor = torch.tensor([1.0, 2.0], device=device_type, requires_grad=True)
+    original_tensor = distribute_tensor(original_tensor, device_mesh, [Shard(0)])
+    out = original_tensor.sum()
+    if plan == 1:
+        out = DTensor.from_local(
+            out.to_local(grad_placements=[Replicate()]),
```
Comment (wconstab) [inline test.py]: maybe add a comment in the test that grad_placements=Replicate is only added for explicitness but the default behavior would have been the same? (or did i misunderstand)

File: torch/distributed/tensor/_api.py [hunk 3]
```diff
@@ -97,7 +97,16 @@ def backward(ctx, grad_output: torch.Tensor):  # type: ignore[override]
             grad_output, mesh, dtensor_spec.placements
         )
         tensor_stride = tuple(tensor_stride)
-        grad_placements = grad_placements or dtensor_spec.placements
+
+        if grad_placements is None:
```
Comment (wconstab) [inline torch/distributed/tensor/_api.py:122]: maybe make this a helper fn and add a # see note normalized grad placements comment to direct readers to where you already documented this behavior previously

File: test.py (hunk 1 above)
Comment (acisseJZhong) [reply test.py]: sorry let me repurpose this to a unit test, this was just my local testing.

File: test.py (hunk 2 above)
Comment (acisseJZhong) [reply test.py]: yes, let me remove this grad_placement, as it should work with the fix. And leave a comment suggesting that user should provide `grad_placement` at their best.

**Review by wconstab — APPROVED** [review]

**Comment by acisseJZhong** [comment]
@pytorchmergebot merge -i 

the failing inductor test is failing across all recent PRs https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50
